src/python/pants/engine/internals/selectors.py

```diff
@@ -117,9 +198,31 @@ def __str__(self) -> str:
         return repr(self)
 
 
+_Awaitable = TypeVar("_Awaitable")
+
+
+class _AwaitableInitTrampoline(Generic[_Awaitable, _Output, _Input]):
```

Comment:
Unclear to me what the types here should even be.

I am getting this error when even trying to run `fmt` on this code:

```
Traceback (most recent call last):
  File "/Users/tdyas/TC/pants/src/python/pants/bin/pants_loader.py", line 19, in <module>
    from pants.bin.pants_runner import PantsRunner
  File "/Users/tdyas/TC/pants/src/python/pants/bin/pants_runner.py", line 13, in <module>
    from pants.bin.remote_pants_runner import RemotePantsRunner
  File "/Users/tdyas/TC/pants/src/python/pants/bin/remote_pants_runner.py", line 15, in <module>
    from pants.option.global_options import GlobalOptions
  File "/Users/tdyas/TC/pants/src/python/pants/option/global_options.py", line 27, in <module>
    from pants.engine.environment import CompleteEnvironment
  File "/Users/tdyas/TC/pants/src/python/pants/engine/environment.py", line 10, in <module>
    from pants.engine.rules import collect_rules, rule
  File "/Users/tdyas/TC/pants/src/python/pants/engine/rules.py", line 29, in <module>
    from pants.engine.goal import Goal
  File "/Users/tdyas/TC/pants/src/python/pants/engine/goal.py", line 12, in <module>
    from pants.option.subsystem import Subsystem
  File "/Users/tdyas/TC/pants/src/python/pants/option/subsystem.py", line 12, in <module>
    from pants.engine.internals.selectors import AwaitableConstraints, Get
  File "/Users/tdyas/TC/pants/src/python/pants/engine/internals/selectors.py", line 323, in <module>
    class Effect(Generic[_Output, _Input], Awaitable[_Output, _Input]):
TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
```

cc @Eric-Arellano